streams/runners/storm/sampling/src/main/java/com/hortonworks/streamline/streams/sampling/service/storm/StormTopologySamplingService.java

```diff
@@ -0,0 +1,257 @@
+        return client.disableSampling(topologyId, asUser);
+    }
+
+    @Override
+    public boolean disableSampling(Topology topology, TopologyComponent component, String asUser) {
+        String topologyId = StormTopologyUtil.findStormTopologyId(client, topology.getId(), asUser);
+        return client.disableSampling(topologyId, component.getId() + "-" + component.getName(), asUser);
+    }
+
+    @Override
+    public SamplingStatus getSamplingStatus(Topology topology, String asUser) {
+        String topologyId = StormTopologyUtil.findStormTopologyId(client, topology.getId(), asUser);
+        return buildSamplingStatus(client.getSamplingStatus(topologyId, asUser));
+    }
+
+    @Override
+    public SamplingStatus getSamplingStatus(Topology topology, TopologyComponent component, String asUser) {
+        String topologyId = StormTopologyUtil.findStormTopologyId(client, topology.getId(), asUser);
+        return buildSamplingStatus(client.getSamplingStatus(topologyId, asUser));
+    }
+
+    @Override
+    public SampledEvents getSampledEvents(Topology topology, TopologyComponent component, EventQueryParams qps, String asUser) {
+        List<SampledEvent> events = new ArrayList<>(qps.count());
+        String topologyId = StormTopologyUtil.findStormTopologyId(client, topology.getId(), asUser);
+        String componentId = component.getId() + "-" + component.getName();
+        Long cur;
+        if (qps.desc()) {
+            cur = qps.start();
+        } else {
+            cur = qps.start() == null ? Long.valueOf(0) : qps.start();
+        }
+        Long nextOffset;
+        Long prevOffset = null;
+        do {
+            String res = client.getSampledEvents(topologyId, componentId, asUser, cur, qps.length());
+            Matcher match = LOG_CONTENT_REGEX.matcher(res);
+            if (!match.find()) {
+                break;
+            }
+            String lines = match.group(1);
+            if (!StringUtils.isEmpty(lines)) {
+                addSampledEvents(lines, componentId, events);
+            }
+            match = PREV_START_REGEX.matcher(res);
+            prevOffset = match.find() ? Long.parseLong(match.group(1)) : null;
+            match = NEXT_START_REGEX.matcher(res);
+            nextOffset = match.find() ? Long.parseLong(match.group(1)) : null;
+            cur = qps.desc() ? prevOffset : nextOffset;
+        } while (events.size() < qps.count() && cur != null);
+
+        long next = 0;
+        Integer length = null;
+        if (!events.isEmpty()) {
+            if (events.size() > qps.count()) {
+                int i = qps.desc() ? events.size() - qps.count() : 0;
+                int j = qps.desc() ? events.size() : qps.count();
+                events = events.subList(i, j);
+            }
+            if (qps.desc()) {
+                if (qps.start() != null) {
+                    next = qps.start() + events.get(0).getStartOffset() - qps.length() + 1;
```

Comment:
This doesn't hold true with fetching multiple pages. You may need to play with `cur` instead of `qps.start()`, or modify `addSampledEvents()` to ensure each event has absolute offset rather than relative offset in page. IMHO latter way is clearer.

Reply:
If we could make event having absolute offset, I don't think we need to calculate with prevOffset too. And the case for calculating `length` should be simplified too, cause we just need to deal with `next` and `qps.length()`: `qps.start()` no needed.

Reply:
for now i can play with `cur`. During the initial query (without any start and length) the start offset is unknown, so would be difficult to find the absolute offset.